[Core] Dynamic image size support for VLMs

DarkLight1337 opened the pull request:
This PR uses the input registry introduced by #5214 to implement an input process that inserts image tokens automatically at the `LLMEngine` level, so that it applies to `LLM.generate`.

Accordingly, I have updated LLaVA-NeXT and Phi-3-Vision to support dynamic image size. Along the way, I have expanded the VLM tests to consider text-only and multiscale-image input in addition to the current single-scale image input.

Based on this, I have written a [detailed guide](https://vllm--5276.org.readthedocs.build/en/5276/dev/multimodal/adding_multimodal_model.html) on how to implement multimodal vLLM models.

Please note that this introduces a **breaking change** to users. Instead of manually repeating image tokens, the same prompt format as described in the corresponding HuggingFace repo should be used regardless of the model.

### Related contributions

Follow-up to #5214.

This PR conflicts with #5237 as it inserts image tokens at the `OpenAIServing` level. This PR has removed such logic from the server to avoid double insertion.

- #5214
- #5237

DarkLight1337 committed:
Introduce a higher level `INPUT_REGISTRY`

DarkLight1337 committed:
Move dummy data generation to input registry

DarkLight1337 committed:
Update docs

DarkLight1337 committed:
Rename `process_input` to `map_input`

DarkLight1337 committed:
Reorder arguments

DarkLight1337 committed:
Apply input processor

DarkLight1337 committed:
Remove `VisionLanguageConfig` from input mapper

DarkLight1337 committed:
Fix bad use of `functools.partial`

DarkLight1337 committed:
Use default input processor

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Fix wrong arguments

DarkLight1337 committed:
Use pillow image instead of tensor to avoid bypassing the processor by default

DarkLight1337 committed:
Update interface of dummy data factory and input processor

DarkLight1337 committed:
Use `InputContext` to handle checked type cast of config types

DarkLight1337 committed:
Add input processor for injecting image tokens; fix docs

DarkLight1337 committed:
Add new documentation pages

DarkLight1337 committed:
Fix LLaVA-NeXT input processor and cleanup code

DarkLight1337 committed:
Fix LLaVA-NeXT input processor and cleanup code

DarkLight1337 committed:
Add sanity check

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Update LLaVA-NeXT

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Update name

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Update LLaVA-NeXT

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Remove `MULTIMODAL` convenience property as it was causing some (import-related?) problems with ray workers

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Update docs

DarkLight1337 committed:
Remove double processing of image tokens

DarkLight1337 committed:
Add docs

DarkLight1337 committed:
Add docs

DarkLight1337 committed:
Add docs

DarkLight1337 committed:
Add docs

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Remove more instances of double processing; update docs

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Remove xfail

DarkLight1337 committed:
Fix missing image token in OpenAI API serving

DarkLight1337 committed:
Fix LLaVA-NeXT test

DarkLight1337 committed:
Remove duplicate processing in async engine

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Set up dummy data factory for phi3v

DarkLight1337 committed:
Move dummy data factories to model files

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Move input processors to model files

DarkLight1337 committed:
Set up input processor for phi3v

DarkLight1337 committed:
Fix wrong feature size

DarkLight1337 committed:
Fix wrong feature size

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Fix wrong feature size

xwjiang2010 commented on vllm/inputs/registry.py:
noob question, do we want to scope this InputRegistry to only apply to MultiModal case?
In other words, do we foresee it being used for non MultiModal case? and what are those?
@DarkLight1337 @ywang96

DarkLight1337 replied:
It may be used for other additional model inputs, such as passing text embeddings directly to the model in the future.

xwjiang2010 replied:
got it, thanks!

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Update validation

xwjiang2010 commented on tests/models/test_llava_next.py:
```diff
@@ -96,11 +93,6 @@ def test_models(hf_runner, vllm_runner, hf_images, vllm_images,
                                               max_tokens,
                                               images=hf_images)
```
For Llava 1.6, we need to test images with different resolutions (that result in different number of patches, e.g. 2 v.s. 5)
This scenario fails with current implementation.
failing at https://github.com/vllm-project/vllm/blob/c2462129521a64b62ace77b28641d2e3bec5831c/vllm/worker/cpu_model_runner.py#L158

1st image has pixel_values of shape (1, **2**, 3, 336, 336)
2nd image has pixel_values of shape (1, **5**, 3, 336, 336)

They can't be concatenated naively. The resulting Tensor should be of shape (1, 7, 3, 336, 336). Other changes I did: Add a "splits" Tensor ([2, 5]) that needs to be passed as kwargs to llava_next.forward so that the concatenated image features can be split accordingly.

DarkLight1337 replied:
Feel free to commit directly to my fork to fix the problem.

DarkLight1337 replied:
I have fixed this issue. However, I found that the model now deviates slightly from the HF implementation. @ywang96 might this be caused by the new `CLIPVisionModel`?

ywang96 replied:
Hmm... could you try swapping it back to `CLIPVisionModel` from transformers and see if that's what's causing the difference?

DarkLight1337 replied:
After some digging, I found that the `pixel_values` passed to HF model and vLLM model are very slightly different. So it has nothing to do with `CLIPVisionModel`.

DarkLight1337 replied:
There seems to be some precision error going on. In HF, the model takes `torch.float32` `pixel_values` even though the model is in `torch.float16`. I'll update the image processor in vLLM to refrain from changing the dtype then.

DarkLight1337 replied:
The test is still failing after fixing this, so that's not all...

DarkLight1337 replied:
Seems that it is indeed a discrepancy introduced by `CLIPVisionTower`.

Using the debugger on the first test iteration in `python -m pytest tests/models/test_llava_next.py -x -s`:

Value of `selected_image_feature` in
https://github.com/huggingface/transformers/blob/ab0f050b42d903f34d6eb97f3f8c0c07f0517ad2/src/transformers/models/llava_next/modeling_llava_next.py#L786
```
tensor([[[ 1.0977,  0.2158,  0.2979,  ...,  1.6602, -0.3267,  1.0205],
         [ 1.3760,  0.1354,  0.0698,  ...,  1.4639,  0.0743,  0.5303],
         [ 0.7861,  0.1392, -0.3694,  ...,  1.5039, -0.1001, -0.4326],
         ...,
         [ 0.5264,  0.0317,  0.3013,  ...,  1.0049,  0.4236,  0.1664],
         [ 0.4185,  0.1843,  0.3418,  ...,  0.7183,  0.6943,  0.2329],
         [ 0.6025,  0.4297,  0.2101,  ...,  0.9277,  0.3252, -0.1746]],

        [[ 0.6880, -0.3057,  0.0807,  ...,  0.3076,  1.5127,  0.8481],
         [ 0.5137, -0.5161, -0.0663,  ...,  0.3184,  1.6836,  0.6689],
         [ 0.6606,  0.5103, -0.3066,  ...,  1.5156,  0.3296,  1.2295],
         ...,
         [ 0.2827,  0.1671,  0.9004,  ...,  0.2566,  0.2463,  0.7812],
         [ 0.4097,  0.0662,  0.6396,  ...,  1.0625,  0.3174,  0.6572],
         [ 0.4282,  0.0069,  0.7041,  ...,  0.9033,  0.6396, -0.0066]],

        [[ 0.3181,  0.0100,  0.5967,  ..., -0.0105, -0.6138,  0.2446],
         [ 0.3816,  0.4092, -0.1753,  ..., -0.1832,  0.0361,  0.1544],
         [ 0.6299, -0.5400, -0.0608,  ..., -0.3350,  1.4229,  1.0791],
         ...,
         [ 0.2058,  0.0090,  0.0802,  ...,  0.2004,  1.2285,  0.6104],
         [ 0.3054, -0.1927, -0.1033,  ...,  0.0706,  1.0908,  1.0615],
         [ 0.4011, -0.2283,  0.1530,  ..., -0.1183,  1.4502,  0.7383]]],
       device='cuda:0', dtype=torch.float16)
```

Value of `stacked_image_features[:3]` in
https://github.com/vllm-project/vllm/blob/784af1a7f3d72ec498373af94e9db5a9a611210d/vllm/model_executor/models/llava_next.py#L392-L393
```
tensor([[[ 1.0996e+00,  2.1753e-01,  2.9932e-01,  ...,  1.6621e+00,
          -3.2715e-01,  1.0215e+00],
         [ 1.3740e+00,  1.3843e-01,  7.0068e-02,  ...,  1.4619e+00,
           7.5562e-02,  5.2783e-01],
         [ 7.8613e-01,  1.3684e-01, -3.7109e-01,  ...,  1.5059e+00,
          -1.0168e-01, -4.3408e-01],
         ...,
         [ 5.2930e-01,  3.3569e-02,  3.0615e-01,  ...,  1.0010e+00,
           4.2529e-01,  1.6077e-01],
         [ 4.1577e-01,  1.8311e-01,  3.4082e-01,  ...,  7.1924e-01,
           6.9336e-01,  2.2864e-01],
         [ 6.0303e-01,  4.3188e-01,  2.1338e-01,  ...,  9.2676e-01,
           3.2910e-01, -1.7847e-01]],

        [[ 6.8750e-01, -3.0444e-01,  8.2397e-02,  ...,  3.0957e-01,
           1.5137e+00,  8.4863e-01],
         [ 5.1367e-01, -5.2148e-01, -6.6040e-02,  ...,  3.1665e-01,
           1.6826e+00,  6.6455e-01],
         [ 6.6064e-01,  5.0391e-01, -3.0420e-01,  ...,  1.5137e+00,
           3.2910e-01,  1.2275e+00],
         ...,
         [ 2.8491e-01,  1.6528e-01,  8.9941e-01,  ...,  2.5757e-01,
           2.4609e-01,  7.8076e-01],
         [ 4.1211e-01,  6.3416e-02,  6.3818e-01,  ...,  1.0635e+00,
           3.1738e-01,  6.5332e-01],
         [ 4.2578e-01,  6.1035e-03,  7.0459e-01,  ...,  9.0088e-01,
           6.4600e-01, -7.3242e-04]],

        [[ 3.1738e-01,  7.2327e-03,  5.9424e-01,  ..., -6.9580e-03,
          -6.0986e-01,  2.4023e-01],
         [ 3.7695e-01,  4.0625e-01, -1.7358e-01,  ..., -1.9067e-01,
           3.8940e-02,  1.5747e-01],
         [ 6.2646e-01, -5.3613e-01, -6.1279e-02,  ..., -3.3447e-01,
           1.4268e+00,  1.0752e+00],
         ...,
         [ 2.0984e-01, -6.7139e-04,  7.8125e-02,  ...,  2.0508e-01,
           1.2275e+00,  6.1328e-01],
         [ 3.0664e-01, -1.8726e-01, -1.0083e-01,  ...,  7.2998e-02,
           1.0928e+00,  1.0674e+00],
         [ 4.0601e-01, -2.3083e-01,  1.4941e-01,  ..., -1.1523e-01,
           1.4512e+00,  7.3926e-01]]], device='cuda:0', dtype=torch.float16)
```

DarkLight1337 replied:
After fixing the issue with floating-point precision, I have reverted the prompts to exactly the HF format (image after user) and now the tests pass.

DarkLight1337 replied:
For now, I've marked the test as `xfail` and updated the test pipeline to show the best `max_tokens` that passes thte test.

DarkLight1337 committed:
Fix image feature calculation for phi3v

DarkLight1337 committed:
Remove redundant code

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Apply isort

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Apply yapf

DarkLight1337 committed:
Reduce `max_tokens` so that test still passes

DarkLight1337 committed:
Fix vllm to hf output (+ rename)

DarkLight1337 committed:
Fix wrong arguments

DarkLight1337 committed:
Move `DummyImageDataFactories` into CLIP model file

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Move `input_processor_for_clip` into CLIP

DarkLight1337 committed:
Remove some magic numbers

DarkLight1337 committed:
Test multiscale inputs for LLaVA-NeXT

DarkLight1337 committed:
Handle multiscale inputs (different number of patches per batch) in LLaVA-NeXT

xwjiang2010 committed:
[VLM] Remove support for pixel_values and image_features.

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

DarkLight1337 committed:
Fix wrong feature size

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Revert max_tokens

DarkLight1337 committed:
Add more tests for input mapper

DarkLight1337 committed:
Sanity check: Also test multiscale inputs for LLaVA-1.5
- Remove test for feature size as it will be unsupported after #5852 anyway

DarkLight1337 committed:
Do not auto-convert image dtype to model's dtype
- This is not done in HF, so we follow them such that consistency tests can pass

DarkLight1337 committed:
Update prompts

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Fix mapper tests w.r.t. dtype change

DarkLight1337 committed:
Clarify docs and add todo

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Remove TODO since vision config will be removed soon

DarkLight1337 committed:
Expand docs

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Add ref

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Update docs

DarkLight1337 committed:
Add docs

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Fix name

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Add `MultiModalInputs` to docs

DarkLight1337 committed:
Fix and add links

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Fix `is_multiscale` not provided anymore

DarkLight1337 committed:
Also test multiscale input for phi3v

DarkLight1337 committed:
Revert max_tokens for phi3v as numerical error still persists

DarkLight1337 committed:
Improve error message

DarkLight1337 committed:
Log the full output for easier reference

DarkLight1337 committed:
Update xfail to be more efficient
- Also fix llava test

DarkLight1337 committed:
Also xfail llava test

xwjiang2010 committed:
address comments

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
remove image_input_type altogether.

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
types

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
format

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

DarkLight1337 committed:
Update comment

DarkLight1337 committed:
Update comment

DarkLight1337 committed:
Update docs

DarkLight1337 committed:
Update docs

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Add see also

DarkLight1337 committed:
Update examples prompt format

ywang96 committed:
modify llava_next

DarkLight1337 committed:
Use dynamic image feature size calculation

DarkLight1337 committed:
Fix phi3v not handling `image_sizes` correctly

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Fix config

DarkLight1337 committed:
Fix config

DarkLight1337 committed:
Update docs

DarkLight1337 committed:
Update docs

DarkLight1337 committed:
Fix `MultiModalInputs` not working in Python 3.8

DarkLight1337 committed:
Fix `_ImageAssets` not working in Python 3.8

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

ywang96 committed:
update example

ywang96 committed:
update doc

DarkLight1337 committed:
Merge branch 'mm-image-tokenizer' into mm-image-tokenizer-2

DarkLight1337 committed:
Update docs

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Fix OpenAI server not working for phi3v

DarkLight1337 committed:
Preemptively handle upcoming models

DarkLight1337 commented on vllm/model_executor/models/phi3v.py:
```diff
@@ -272,10 +278,8 @@ def _get_phi3v_image_feature_size(
-
-    #TODO: change the logic for dummy data to support dynamic shape
-    _, _, dummy_height, dummy_width = multimodal_config.image_input_shape
+    # TODO: How to get the max possible feature size?
```
@Isotr0py does the model have a cap on the maximum feature size? If so, can you figure out what that is? Thanks!

Isotr0py replied:
In my test, `Phi3v` has a cap maximum `feature size` with 2473 and `fuyu` is 2304 for single image.

For `Phi3v`, the feature size depends on the image size ratio (`hd_num`) whose max cap is 16:1 by default (correspond to 2473 feature size).

For `fuyu`, all input images will be resized to 1920x1080 by default, and then it will re-slice the resized image to unpadded image size or clamp out of box size to 1920/1080. So a 1920x1080 image will cause maximum feature size  with 2473. (And seems that `chameleon` also use a similar image processing logic with `fuyu`)

Generally, I think most of models should have a cap on the maximum feature size, but it depends on the image processing logic and what's the exact target image size.

DarkLight1337 replied:
Thanks for the quick response! I'll add them into this PR accordingly.

DarkLight1337 committed:
Add more models

DarkLight1337 committed:
Update feature size for dummy data

xwjiang2010 committed:
Merge branch 'main' of https://github.com/vllm-project/vllm into remove_image_features_2

xwjiang2010 committed:
format

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
ExternalMultiModalDataDict

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 commented on tests/models/test_llava_next.py:
```diff
@@ -14,21 +18,21 @@
         (672, 672): 2928,
-        (1344, 336): 1944,
-        (336, 1344): 1890,
+        (1344, 336): 2928,
```
why 2928? How's it calculated?

DarkLight1337 replied:
It's mostly a holdover from the previous code. This is not used anymore but we still need to pass a value to the config until your PR removes it.

xwjiang2010 committed:
mention schema

DarkLight1337 committed:
Use a less strict check

DarkLight1337 committed:
Fix phi3v test

DarkLight1337 committed:
Update default length as the dummy image feature size is increased

DarkLight1337 committed:
Raise full error if output is completely different

DarkLight1337 committed:
Fix phi3v not using input processor

DarkLight1337 committed:
Move size factors outside

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Fix some outputs not being checked

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Also test no image

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Batch by size factors

DarkLight1337 committed:
Factor out xfail code

DarkLight1337 committed:
Fix unused args

ywang96 commented on vllm/model_executor/models/llava_next.py:
```diff
@@ -375,47 +371,54 @@ def _merge_image_patch_embeddings(self, image_size: torch.Tensor,
 
-        return stacked_image_features.view(b, num_patches,
-                                           *stacked_image_features.shape[-2:])
+        return [
+            proj(image_features) for image_features in torch.split(
+                stacked_image_features, num_patches_per_batch)
+        ]
```
Why do we need to make `proj` a variable instead of calling `self.multi_modal_projector` directly? IMO the latter makes the code less convoluted.

DarkLight1337 replied:
Will fix

ywang96 commented on vllm/model_executor/models/phi3v.py:
```diff
@@ -330,27 +300,111 @@ def _calc_hd_transform_size(*, width: int, height: int, hd_num: int = 16):
+    prompt_token_ids = llm_inputs["prompt_token_ids"]
+    tokenizer = cached_get_tokenizer(model_config.tokenizer)
+
+    # We need to get the token for "<", not "▁<"
+    # https://huggingface.co/microsoft/Phi-3-vision-128k-instruct/raw/main/tokenizer.json
+    a_token_id, = tokenizer.encode("a", add_special_tokens=False)
+    a_token_id_, *image_1_token_ids = tokenizer.encode(
+        "a<|image_1|>", add_special_tokens=False)
+    assert a_token_id == a_token_id_
```
IMO we should just hardcode `image_1_token_ids` and leave a NOTE on it. The use of `<image_1>` is such a hack in the original model example.

DarkLight1337 replied:
I left this code here so I can remember how to reconstruct the token representation for additional image tokens when we implement support for #5820 in the future.

DarkLight1337 replied:
I can move this to a separate function to make it less confusing.

ywang96 commented on vllm/model_executor/models/utils.py:
```diff
@@ -0,0 +1,41 @@
+from vllm.multimodal import BatchedTensors
+
+
+def merge_vision_embeddings(input_ids: torch.Tensor,
+                            inputs_embeds: torch.Tensor,
+                            vision_embeddings: BatchedTensors,
+                            image_token_id: int) -> torch.Tensor:
```
Shouldn't we have this function inside `multimodal/utils` since this is only relevant to VLMs?

Personally I'm also not a huge fan of making this a util function - IMO merging vision embeddings & input embeddings is part of the model implementation (it just happened that all the VLMs share the same behavior now), so keeping all relevant functions in the same place is better for readability.

DarkLight1337 replied:
That's why I placed this function under models directory rather than multimodal directory since it's specific to models only. Do you want me to copy this to each model file individually?

ywang96 replied:
I think having some repetitive code in this case is okay since this is only relevant to those specific VLMs.

My only concern of keeping it this way is that we don't know if this is generalizable to all future VLMs, so in the future we might have some models that use this shared util function and other models specify their individual embedding processing function inside their model.py file, and thus it'll might be confusing to the developers whether they should keep all the functions used in the model implementation within the same file, or specify it in another location.

Either way, this is just my nit and you don't have to necessarily agree with it.

DarkLight1337 replied:
I think we probably have to redesign the framework anyway if we need other ways of merging the embeddings. @xwjiang2010 since you came up with this method in the first place, what are your thoughts on this?

ywang96 commented on vllm/multimodal/image.py:
```diff
@@ -1,19 +1,90 @@
+    if prompt is None:
+        new_prompt = None
+    else:
+        image_token_str = tokenizer.decode(image_token_id)
+        pad_token_str_left = (None if pad_token_left is None else
+                              tokenizer.decode(pad_token_left))
+        pad_token_str_right = (None if pad_token_right is None else
+                               tokenizer.decode(pad_token_right))
+        replacement_str = "".join(
+            repeat_and_pad_token(
+                image_token_str,
+                repeat_count=repeat_count,
+                pad_token_left=pad_token_str_left,
+                pad_token_right=pad_token_str_right,
+            ))
+
+        # The image tokens are removed to be consistent with HuggingFace
+        new_prompt = prompt.replace(image_token_str, replacement_str, 1)
```
I don't think we should change and return the `new_prompt` here - doesn't that mean we are overwriting the original prompt provided by user? Why is changing `prompt_token_ids` itself not sufficient?

DarkLight1337 replied:
It is sufficient by itself, but I am worried that the discrepancy between prompt text and token IDs may cause confusion to developers. If you think it's fine, I can remove this.

ywang96 reviewed (commented):
Great work! I left a few comments and questions as my first round of review!

DarkLight1337 commented on .buildkite/test-pipeline.yaml:
```diff
@@ -126,7 +126,7 @@ steps:
   commands:
     - bash ../.buildkite/download-images.sh
-    - pytest -v -s models -m vlm
+    - pytest -v -s -rx models -m vlm
```
`-rx` in this context does not actually mean `-r` and `-x`, as described [here](https://docs.pytest.org/en/stable/how-to/skipping.html#how-to-use-skip-and-xfail-to-deal-with-tests-that-cannot-succeed) (yeah I know it is confusing).

DarkLight1337 replied:
In any case, this will no longer be necessary as I'm working on using logprob-based tests instead of string-based ones. It's quite a hassle to convert the format of the text prompt from vLLM to HF in the test.

DarkLight1337 commented on vllm/model_executor/models/clip.py:
```diff
@@ -80,6 +84,41 @@ def dummy_feature_data_for_clip(
 
+def input_processor_for_clip(
+    model_config: ModelConfig,
+    multimodal_config: VisionLanguageConfig,
```
Yeah, it's just to keep the interface consistent. It will be removed later by@xwjiang2010.

ywang96 replied:
Sounds good - I will add TODO here (after your current CI runs).

DarkLight1337 replied:
On second thought, I'm removing this now since the other input processors now take `ctx` instead of the individual configs anyway.

DarkLight1337 commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -105,71 +135,37 @@ def _parse_chat_message_content_parts(
-            (image_token_prompt,
-             image_token_str) = vlm_config.get_image_token_text(self.tokenizer)
-
-            # NOTE: If image token string (e.g, <image>) is already present
-            # in the text prompt, we assume it follows the same format required
-            # by the engine.
-            if image_token_str in text_prompt:
-                logger.warning(
-                    "Detected image token string in the text prompt. "
-                    "Skipping prompt formatting.")
-                messages = [
-                    ConversationMessage(role=role, content=text_prompt)
-                ]
```
Alright, I'll add this back in.

DarkLight1337 committed:
Check logprobs instead of xfailing

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Fix different scales not being in the same batch

DarkLight1337 committed:
Apply suggestions from code review

Co-authored-by: Roger Wang <136131678+ywang96@users.noreply.github.com>

DarkLight1337 committed:
Add link

DarkLight1337 committed:
Use `self.multi_modal_projector` directly

DarkLight1337 committed:
Allow users to send image token formatted prompt directly

DarkLight1337 commented on tests/models/test_llava.py:
```diff
@@ -89,46 +91,95 @@ def run_test(
+                                                    images=hf_image_inputs)
+
+    # There may be numeric differences for multiscale images due to
+    # our implementation of CLIPVisionModel
```
I've updated the tests to be logprob-based so the tests should still succeed despite some inaccuracies. Let's wait for the CI to run with your `CLIPVisionModel` and check the results before switching.

DarkLight1337 commented on docs/source/models/vlm.rst:
```diff
@@ -141,5 +137,4 @@ A full code example can be found in `examples/openai_vision_api_client.py <https
 .. note::
-    The prompt formatting with the image token ``<image>`` is not needed when serving VLMs with the API server since the prompt will be 
-    processed automatically by the server.
+    There is no need to format the prompt in the API request when since it will be handled by the server.
```
Could you elaborate? I made this change because I no longer mention the use of `<image>` token outside of the LLaVA examples (since it may not be available for certain models).

ywang96 replied:
My point is that we're using `llava-hf/llava-1.5-7b-hf` as an example for both offline and online in this page, so calling out the unnecessary usage of image token (in this example, `<image>`) makes it clear what "formatting" is actually not needed.

DarkLight1337 replied:
OK, perhaps it'll be more appropriate to put it as a comment inside the code block then.

DarkLight1337 committed:
Factor out the code for placeholder token IDs

DarkLight1337 committed:
Remove `-rx` flag

DarkLight1337 committed:
Fix distributed tests

DarkLight1337 committed:
Fix string mismatch warning

DarkLight1337 committed:
Relax phi3v test; add TODO for llava tests

DarkLight1337 committed:
Fix distributed tests

xwjiang2010 committed:
address comments

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
Merge branch 'main' of https://github.com/vllm-project/vllm into remove_image_features_2

xwjiang2010 committed:
format

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
rm ctx

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Fix distributed test

DarkLight1337 commented:
@youkaichao the distributed tests are failing again. Not sure what I'm doing wrong here, since I've already removed the relevant imports from the top level...

DarkLight1337 committed:
Update docs about prompt formatting

DarkLight1337 committed:
Remove unused parameter

DarkLight1337 committed:
Remove unused import

DarkLight1337 committed:
Fix distributed test

xwjiang2010 committed:
rm ImageData and MultiModalData

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
rm external

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
comments

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
fix dist gpu test.

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
address comments

DarkLight1337 committed:
Further avoid cuda init

DarkLight1337 committed:
Add warnings for repeated image tokens

xwjiang2010 committed:
docs

Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

xwjiang2010 committed:
Update vllm/multimodal/base.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>

xwjiang2010 committed:
format
Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

DarkLight1337 committed:
Reword

DarkLight1337 committed:
Merge branch 'remove_image_features_2' of https://github.com/xwjiang2010/vllm into mm-image-tokenizer-2

DarkLight1337 committed:
Remove useless test

DarkLight1337 committed:
Unify test API between HfRunner and VllmRunner

DarkLight1337 committed:
Fix import error

DarkLight1337 committed:
Fix attribute error

ywang96 committed:
fix import error

ywang96 committed:
update llava next example

DarkLight1337 committed:
Merge branch 'remove_image_features_2' of https://github.com/xwjiang2010/vllm into mm-image-tokenizer-2

DarkLight1337 committed:
Update comments

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Remove some unnecessary deferred imports

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Use more precise type annotation

DarkLight1337 committed:
Fix wrong feature size

DarkLight1337 committed:
Fix wrong image

DarkLight1337 committed:
Remove unnecessary lazy import

DarkLight1337 committed:
Check for conflicting kwargs in `map_input`

DarkLight1337 committed:
Avoid unnecessary processing

DarkLight1337 committed:
Update doc

DarkLight1337 committed:
Avoid cuda init

DarkLight1337 committed:
Remove unused logger

DarkLight1337 committed:
Remove unnecessary deferred imports

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Fix typo

ywang96 commented on examples/llava_next_example.py:
```diff
@@ -5,22 +5,20 @@
-# a fixed image input shape and its corresponding feature size is required.
-# See https://github.com/vllm-project/vllm/pull/4199 for the complete
-# configuration matrix.
+# The assets are located at `s3://air-example-data-2/vllm_opensource_llava/`.
+# You can use `.buildkite/download-images.sh` to download them
```
I actually didn't use those assets for this example

DarkLight1337 replied:
Oops. I'll remove it then.

ywang96 commented on vllm/multimodal/base.py:
```diff
@@ -1,23 +1,108 @@
 
-N = TypeVar("N", bound=Type["nn.Module"])
+
+class MultiModalData:
+    """
+    Base class that contains multi-modal data.
+
+    To add a new modality, add a new file under ``multimodal`` directory.
+
+    In this new file, subclass :class:`~MultiModalData` and
+    :class:`~MultiModalPlugin`.
+
+    Finally, register the new plugin to
+    :const:`vllm.multimodal.MULTIMODAL_REGISTRY`.
+    This enables models to call :meth:`MultiModalRegistry.map_input` for
+    the new modality.
+    """
+    pass
```
I don't see where we use this?

DarkLight1337 replied:
Looks like I missed this when merging the other PR, will remove this. Thanks for the catch!

ywang96 commented on vllm/multimodal/image.py:
```diff
@@ -1,19 +1,101 @@
+            ))
+
+        image_token_count = prompt.count(image_token_str)
+        if image_token_count > 16:
```
Why 16?

DarkLight1337 replied:
I chose this arbitrary number to distinguish between the case where user wants to input multiple images in the same patch, and the case where user is using the old way of repeating image tokens.

ywang96 replied:
Let's leave a NOTE on this then for clarification

DarkLight1337 replied:
Done!

ywang96 reviewed (commented):
Overall LGTM - I'll just need to run some testing on my end before finally approving this!

DarkLight1337 committed:
Address comments

DarkLight1337 committed:
Add comment

ywang96 committed:
Merge branch 'main' into mm-image-tokenizer-2

DarkLight1337 committed:
Merge branch 'upstream' into mm-image-tokenizer-2

DarkLight1337 committed:
Update XPU runner's multimodal logic

DarkLight1337 committed:
Fix unused import

DarkLight1337 committed:
Fix feature size calculation

DarkLight1337 committed:
Add extra image to test

DarkLight1337 committed:
Support multimodal data for neuron and tpu

DarkLight1337 committed:
Fix broadcasting

DarkLight1337 committed:
Fix OpenVINO model runner for multimodal data

DarkLight1337 committed:
Cleanup

ywang96 reviewed (approved):
LGTM! Thank you for the work and glad we resolved all the issues!
